### PR TITLE
Correctly resume from after external calls.

### PR DESCRIPTION
**Pull request by vext01**
**This is a draft**

Before we would run the block containing the call twice, which is
incorrect. This is the reason ykcbf printed everything twice.

Only a draft for now because I think I can do some minor refactoring to make the code neater. I wanted to share the code in case it helps some of Lukas' Ruby issues.

[We will also need to do some [quite major refactoring](https://github.com/ykjit/yk/issues/385) in the medium-term]

### CODE REVIEW

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 1]
```diff
@@ -69,6 +69,8 @@ class JITModBuilder {
     }
   }
 
+  // FIXME: this function need to be refactored.
```
Comment (vext01) [inline ykllvmwrap/src/jitmodbuilder.cc:72]: Gah, need**s*

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 2]
```diff
@@ -139,22 +142,34 @@ class JITModBuilder {
 
       // Iterate over all instructions within this block and copy them over
       // to our new module.
+      // FIXME It would be nice to merge the two skipping mechanisms
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:145]: `FIXME` or `FIXME:` (I'm indifferent, but we should be consistent).

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 3]
```diff
@@ -139,22 +142,34 @@ class JITModBuilder {
...
+          }
+        }
 
-        // If we've returned from a call, skip ahead to the instruction where
-        // we left off.
+        // If we've returned from an internal call, skip ahead to the
+        // instruction where we left off.
         if (last_call != nullptr) {
           if (&*I == last_call) {
             last_call = nullptr;
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:162]: Silly question perhaps: does this deal with arbitrary levels of "if we've returned from an internal call" due to inlining?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 4]
```diff
@@ -176,10 +191,16 @@ class JITModBuilder {
                   CF->getName(), JITMod);
               VMap[CF] = DeclFunc;
             }
+            copyInstruction(&Builder, (Instruction *)&*I);
             // We should expect an "unmappable hole" in the trace. This is
             // where the trace followed a call into external code for which be
             // have no IR, and thus we cannot map blocks for.
             ExpectUnmappable = true;
+            // Peek at the instruction we will need to resume at.
+            auto PeekIter = I;
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:200]: Aren't these three lines the same as `ExcCallInst = &*(I++);` ?

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:162]: Yes. Lukas implemented that. See the fib test.

Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:162]: The thing that worries me is that `last_call` suggests we can only handle this to one level. Perhaps that's all we need to handle, but it's not obvious to me.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 4 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:200]: `PeekIter = I` makes a copy of the iterator, leaving the original iterator untouched. Apparently this is how you peek an iterator in C++: copy the iterator.

Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:200]: My eyes hurt! But I have no idea about this stuff, so I shall not rant further :)

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:162]: `last_call` is populated by popping from a stack of inlined calls.

But @ptersilie is your guy for this logic. I'm still a little over-whelmed by it!

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 4 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:200]: It's a C++ thing. When you have a value on the stack, and assign something from it, you call a "copy constructor" to make a fresh instance.

It tripped me too when I first started with C++.

Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:200]: Just to be 100% sure:

```
#include <iostream>
#include <vector>

using namespace std;

int main(void) {
    vector<int> MyVec = {1, 2, 3};
    auto it1 = MyVec.begin();
    auto it2 = it1;
    it2 += 2;
    cout << "it1: " << *it1 << "\n";
    cout << "it2: " << *it2 << "\n";
}
```

Gives:
```
it1: 1
it2: 3
```

**Comment by vext01** [comment]
That should be all comments addressed.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:162]: Can you file an issue / talk to @ptersilie about this? I agree that it's not something for this PR, but it does look fishy, and I don't want us to forget to look at it.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
Sorry, one more change. A variable name was misleading. OK to squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 3 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:162]: `last_call` is populated here:
```
if (isa<ReturnInst>(I)) {
  last_call = inlined_calls.back();
}
```
where `inlined_calls` is a vector of calls we've followed (and the tracer has inlined).

**Pull request merged**